image-mdraid.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2025 Tomas Mudrunka <harviecz@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+
+MDRAID Superblock generator
+This should create valid mdraid superblock for raid1 with 1 device (more devices can be added once mounted).
+It is still very basic, but following seems to be working:
+
+mdadm --examine test.img
+losetup /dev/loop1 test.img
+mdadm --assemble md /dev/loop1
+
+Some docs:
+https://raid.wiki.kernel.org/index.php/RAID_superblock_formats#Sub-versions_of_the_version-1_superblock
+https://docs.huihoo.com/doxygen/linux/kernel/3.7/md__p_8h_source.html
+
+*/
+
+#include <confuse.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <linux/raid/md_p.h>
+
+#include "genimage.h"
+
+#define DATA_OFFSET_SECTORS	(2048)
+#define DATA_OFFSET_BYTES	(DATA_OFFSET_SECTORS*512)
+
+static void random_uuid(__u8 *buf)
+{
+	__u32 r[4];
+	for (int i = 0; i < 4; i++)
+		r[i] = random();
+	memcpy(buf, r, 16);
+}
+
+static unsigned int calc_sb_1_csum(struct mdp_superblock_1 * sb)
+{
+	unsigned int disk_csum, csum;
+	unsigned long long newcsum;
+	int size = sizeof(*sb) + __le32_to_cpu(sb->max_dev)*2;
+	unsigned int *isuper = (unsigned int*)sb;
+
+	/* make sure I can count... (needs include cstddef) */
+	/*
+	if (offsetof(struct mdp_superblock_1,data_offset) != 128 ||
+	    offsetof(struct mdp_superblock_1, utime) != 192 ||
+	    sizeof(struct mdp_superblock_1) != 256) {
+		fprintf(stderr, "WARNING - superblock isn't sized correctly\n");
+	}
+	*/
+
+	disk_csum = sb->sb_csum;
+	sb->sb_csum = 0;
+	newcsum = 0;
+	for (; size>=4; size -= 4 ) {
+		newcsum += __le32_to_cpu(*isuper);
+		isuper++;
+	}
+
+	if (size == 2)
+		newcsum += __le16_to_cpu(*(unsigned short*) isuper);
+
+	csum = (newcsum & 0xffffffff) + (newcsum >> 32);
+	sb->sb_csum = disk_csum;
```

Comment:
This is redundant. `sb_csum` is immediately overwritten by the return value of this function. Just set the new value directly here.

Reply:
not really, calling the function does not imply you want to set the checksum in the metadata. if you read carefuly you will see it temporarily sets the checksum to 0 and then sets it back to original value (rather than calculated one). This is done because checksum in metada has to be 0 when calculating checksum of metadata, otherwise that's gonna be endless recursion.